combinetf2/fitter.py

```diff
@@ -1058,42 +1144,44 @@ def _compute_yields(self, inclusive=True, profile=True, full=True):
             return normcentral
 
     @tf.function
-    def expected_with_variance(
-        self,
-        fun,
-        profile=False,
-        compute_cov=False,
-        compute_global_impacts=False,
-        inclusive=True,
-        full=True,
-        need_observables=True,
-    ):
-        return self._expvar(
-            fun_exp=fun,
-            compute_cov=compute_cov,
-            compute_global_impacts=compute_global_impacts,
-            profile=profile,
-            inclusive=inclusive,
-            full=full,
-            need_observables=need_observables,
-        )
+    def expected_with_variance(self, *args, **kwargs):
+        return self._expected_with_variance(*args, **kwargs)
 
     @tf.function
-    def expected_variations(
-        self,
-        fun,
-        correlations=False,
-        inclusive=True,
-        full=True,
-        need_observables=True,
-    ):
-        return self._expvariations(
-            fun,
-            correlations=correlations,
-            inclusive=inclusive,
-            full=full,
-            need_observables=need_observables,
-        )
+    def expected_variations(self, *args, **kwagrs):
+        return self._expected_variations(*args, **kwagrs)
+
+    def _chi2(self, res, res_cov, ndf_reduction=0):
+        res = tf.reshape(res, (-1, 1))
+        ndf = tf.size(res) - ndf_reduction
+
+        if ndf_reduction > 0:
+            # covariance matrix is in general non invertible with ndf < n
+            # compute chi2 using pseudo inverse
+            chi_square_value = tf.transpose(res) @ tf.linalg.pinv(res_cov) @ res
+        else:
+            chi_square_value = tf.transpose(res) @ tf.linalg.solve(res_cov, res)
+
+        return tf.squeeze(chi_square_value), ndf
+
+    @tf.function
+    def chi2(self, model, profile):
```

Comment:
It's not recommended to pass python objects (model) as parameters to tf.functions

At best it will result in retracing every single time, at worst incorrect/undefined behaviour.